// ticketops-frontend/src/constants/permissions.js
export const PERMISSIONS = {
    CREATE_TICKET: 'CREATE_TICKET',
    EDIT_TICKET: 'EDIT_TICKET',
    DELETE_TICKET: 'DELETE_TICKET',
    VIEW_REPORTS: 'VIEW_REPORTS',
    MANAGE_USERS: 'MANAGE_USERS',
    MANAGE_SITES: 'MANAGE_SITES',
    MANAGE_ASSETS: 'MANAGE_ASSETS',
    ESCALATION_L1: 'ESCALATION_L1',
    ESCALATION_L2: 'ESCALATION_L2',
    ESCALATION_L3: 'ESCALATION_L3',
    DIRECT_STOCK_REPLACEMENT: 'DIRECT_STOCK_REPLACEMENT',
    DIRECT_RMA_GENERATE: 'DIRECT_RMA_GENERATE',
    MANAGE_SITE_STOCK: 'MANAGE_SITE_STOCK'
};

export const PERMISSION_LABELS = {
    [PERMISSIONS.CREATE_TICKET]: 'Create Tickets',
    [PERMISSIONS.EDIT_TICKET]: 'Edit Tickets',
    [PERMISSIONS.DELETE_TICKET]: 'Delete Tickets',
    [PERMISSIONS.VIEW_REPORTS]: 'View Reports',
    [PERMISSIONS.MANAGE_USERS]: 'Manage Users',
    [PERMISSIONS.MANAGE_SITES]: 'Manage Sites',
    [PERMISSIONS.MANAGE_ASSETS]: 'Manage Assets',
    [PERMISSIONS.ESCALATION_L1]: 'Escalation Team Level 1',
    [PERMISSIONS.ESCALATION_L2]: 'Escalation Team Level 2',
    [PERMISSIONS.ESCALATION_L3]: 'Escalation Team Level 3',
    [PERMISSIONS.DIRECT_STOCK_REPLACEMENT]: 'Direct Stock Replacement',
    [PERMISSIONS.DIRECT_RMA_GENERATE]: 'Direct RMA Generation',
    [PERMISSIONS.MANAGE_SITE_STOCK]: 'Manage Site Stock'
};
